File: snapmaker-u1/scripts/em-dash-guard.mjs
// RULE ZERO guard: fail if any authored adapter file contains an em-dash or en-dash. No linter
// enforces this, so this guard walks the adapter (client TS + jinni Python + scripts) and exits
// non-zero on any offender. The dash codepoints are written as escapes so the guard never trips on
// itself. Mirrors the daemon's em_dash_guard.py, which is daemon-rooted and does not scan .ts.
import { readFileSync, readdirSync } from 'node:fs'
import { join, extname } from 'node:path'
import { fileURLToPath } from 'node:url'

const ADAPTER_ROOT = join(fileURLToPath(new URL('.', import.meta.url)), '..')
const EM_DASH = String.fromCharCode(0x2014)
const EN_DASH = String.fromCharCode(0x2013)
const SCANNED = new Set(['.ts', '.py', '.sh', '.mjs', '.json', '.md'])
const EXCLUDED_DIRS = new Set([
  'node_modules', '.venv', '__pycache__', '.pytest_cache', '.mypy_cache', '.ruff_cache', '.git',
])

function sourceFiles(dir) {
  return readdirSync(dir, { withFileTypes: true }).flatMap((entry) => {
    if (EXCLUDED_DIRS.has(entry.name)) return []
    const full = join(dir, entry.name)
    if (entry.isDirectory()) return sourceFiles(full)
    return SCANNED.has(extname(entry.name)) ? [full] : []
  })
}

const offenders = sourceFiles(ADAPTER_ROOT).filter((file) => {
  const text = readFileSync(file, 'utf8')
  return text.includes(EM_DASH) || text.includes(EN_DASH)
})

if (offenders.length > 0) {
  console.error('Banned em-dash / en-dash found in:')
  for (const file of offenders) console.error(`  ${file}`)
  process.exit(1)
}

File: klipper-jinni/scripts/em_dash_guard.py
#!/usr/bin/env python3
"""RULE ZERO guard for the klipper-jinni app: fail on any em-dash (U+2014) or en-dash (U+2013).

Banned everywhere in Bespok3d; no linter enforces it. Walks this app's authored text by suffix,
skipping caches and build output. The dash codepoints are written as escapes so the guard never
trips on itself.
"""
import sys
from pathlib import Path

EM_DASH = chr(0x2014)
EN_DASH = chr(0x2013)
SCANNED_SUFFIXES = (".py", ".md", ".json", ".sh", ".toml", ".txt")
EXCLUDED_DIRS = (
    ".venv", ".git", "__pycache__", ".mypy_cache", ".pytest_cache", ".ruff_cache", ".hypothesis",
)


def is_scanned(path: Path) -> bool:
    if not path.is_file() or any(part in EXCLUDED_DIRS for part in path.parts):
        return False
    return path.suffix in SCANNED_SUFFIXES


def has_banned_dash(path: Path) -> bool:
    text = path.read_text(encoding="utf-8", errors="ignore")
    return EM_DASH in text or EN_DASH in text


def main() -> int:
    root = Path(__file__).resolve().parent.parent
    offenders = [path for path in root.rglob("*") if is_scanned(path) and has_banned_dash(path)]
    for path in offenders:
        print(f"RULE ZERO violation (em-dash/en-dash): {path.relative_to(root)}")
    return 1 if offenders else 0


if __name__ == "__main__":
    sys.exit(main())
